parser.y

```diff
@@ -5303,11 +5306,21 @@ InsertValues:
 	}
 |	'(' ColumnNameListOpt ')' SelectStmt
 	{
-		$$ = &ast.InsertStmt{Columns: $2.([]*ast.ColumnName), Select: $4.(*ast.SelectStmt)}
+		x := &ast.InsertStmt{Columns: $2.([]*ast.ColumnName), Select: $4.(*ast.SelectStmt)}
+		st := $4.(*ast.SelectStmt)
+		if st.SelectStmtOpts.TableHints != nil {
+			x.TableHints = st.SelectStmtOpts.TableHints
+		}
+		$$ = x
 	}
 |	'(' ColumnNameListOpt ')' '(' SelectStmt ')'
 	{
-		$$ = &ast.InsertStmt{Columns: $2.([]*ast.ColumnName), Select: $5.(*ast.SelectStmt)}
+		x := &ast.InsertStmt{Columns: $2.([]*ast.ColumnName), Select: $5.(*ast.SelectStmt)}
+		st := $5.(*ast.SelectStmt)
+		if st.SelectStmtOpts.TableHints != nil {
+			x.TableHints = st.SelectStmtOpts.TableHints
+		}
+		$$ = x
 	}
 |	'(' ColumnNameListOpt ')' UnionStmt
 	{
```

Comment:
also, how about this?

```sql
insert /*+ qb_name(qb1) */ into a 
    select /*+ qb_name(qb2) */ 1 union select /*+ qb_name(qb3) */ 2;
```

Reply:
Do you have any suggestion? I'm not sure how to solve this case properly.

Reply:
i think you should *not* lift the hint from SELECT/UNION into INSERT in the parser. leave everything separate and let TiDB handle the propagation.

Reply:
That's a good idea. We shall only handle the insert hint in its own stmt struct.